[BUG-FIX] Mentions Bug fixed

sarthakpranesh opened the pull request:
1. Updated the regex to look for other mention categories too.
2. gave new meaningful names to the variables
3. Updated the params passed to identifyMentionKeyword()

@RocketChat/ReactNative
Closes #1580 

What was the bug?
In the web site, whenever we used "@" a list of possible mentions would appear whereas this wasn't the case in the mobile app. We had to enter one extra character after "@", "#" and ":" to give the possible mentions.

Is the bug fixed?
YES, now all the symbols "@", "#", ":" and "/" have the expected behavior. Showcased in the below screenshots.

![3](https://user-images.githubusercontent.com/41206172/73969660-da285680-4941-11ea-8e67-c87174113a37.jpg)
![2](https://user-images.githubusercontent.com/41206172/73969662-dc8ab080-4941-11ea-84a3-fd3dbcc7182d.jpg)
![1](https://user-images.githubusercontent.com/41206172/73969671-ddbbdd80-4941-11ea-92a9-3b8372b814f7.jpg)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative?pullRequest=1673) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative?pullRequest=1673) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat.ReactNative?pullRequest=1673) it.</sub>

djorkaeffalexandre commented on app/containers/MessageBox/index.js:
```diff
@@ -292,17 +292,16 @@ class MessageBox extends Component {
+				if (!mentionWithSearch) {
+					const mentionWithOutSearch = lastNativeText.match(/(#|@|:|^\/)$/);
+					if (mentionWithOutSearch) {
+						return this.identifyMentionKeyword('a', mentionWithOutSearch[0]);
```
Hello @sarthakpranesh, very thanks about your contribution.
For first, you solution not is about the issue #1580, but is another problem that we needs to resolve! :)
But, this `'a'` char isn't the better solution to this problem, how you can see [here](https://github.com/sarthakpranesh/Rocket.Chat.ReactNative/blob/Sarthak-MentionFix/app/lib/rocketchat.js#L543), we have a control to turn off promises when you're searching for something, you need to change this logic to consider a blank string in the search, and only add others mentions track on `lastNativeText.match`, it'll fix the problem.

sarthakpranesh replied:
Cool, I'll make the required changes asap :+1:

sarthakpranesh commented:
Updated app/lib/rocketchat.js search function
1. Removed unnecessary if condition which was not allowing the search for the blank string for any mention without any search text
2. Removed 'a' char from this.identifyMentionKeyword function and added blank string

Let me know if there are any further changes that are required to be made for this problem. I would definitely open another PR soon for issue #1580 

Warm Regards

djorkaeffalexandre commented on app/lib/rocketchat.js:
```diff
@@ -540,11 +540,6 @@ const RocketChat = {
 			this.oldPromise('cancel');
 		}
 
-		if (searchText === '') {
```
You can't remove this without add another logic to `delete` the `oldPromise` when search was finished.

sarthakpranesh replied:
Ok, I'll look into it and will see what is required

diegolmello commented on app/containers/MessageBox/index.js:
```diff
@@ -295,17 +295,16 @@ class MessageBox extends Component {
-						return this.identifyMentionKeyword('', MENTIONS_TRACKING_TYPE_COMMANDS);
+				const mentionWithSearch = lastNativeText.substr(0, cursor).match(regexp);
+				if (!mentionWithSearch) {
+					const mentionWithOutSearch = lastNativeText.match(/(#|@|:|^\/)$/);
```
Can you explain what's `mentionWithSearch` and `mentionWithOutSearch`?

sarthakpranesh replied:
mentionWithSearch is basically a variable to check if the characters #, @, : and / are followed by an alphabet or not. If they are followed by an alphabet then we need to separate those alphabets and use it in the search. Whereas the mentionWithOutSearch is to verify if the user just started with @, #, : and / but has entered any alphabet after that, in this case, we just want to make a search with an empty string.

diegolmello commented on app/containers/MessageBox/index.js:
```diff
@@ -295,17 +295,16 @@ class MessageBox extends Component {
-						return this.identifyMentionKeyword('', MENTIONS_TRACKING_TYPE_COMMANDS);
+				const mentionWithSearch = lastNativeText.substr(0, cursor).match(regexp);
+				if (!mentionWithSearch) {
+					const mentionWithOutSearch = lastNativeText.match(/(#|@|:|^\/)$/);
```
Emojis didn't fire with `:`.

sarthakpranesh replied:
I'll look into this asap

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative?pullRequest=1673) <br/>All committers have signed the CLA.

sarthakpranesh commented:
FIX
All mentions ( `@`, `#`, `:`, `/` ) show the expected behavior similar to that of the web client.

Please let me know of any further changes.
![mentions](https://user-images.githubusercontent.com/41206172/75903195-3e6a0780-5e67-11ea-8c57-24e0add566b6.gif)

ankar84 commented:
@sarthakpranesh @diegolmello @djorkaeffalexandre 
Great changes!
Wait them to be merged a lot!
